src/webserver/DOSGuard.h

```diff
@@ -9,11 +9,48 @@
 class DOSGuard
 {
     boost::asio::io_context& ctx_;
-    std::mutex mtx_;  // protects ipFetchCount_
+    std::recursive_mutex
+        mtx_;  // protects ipFetchCount_, ipRequestCount_
+               // recursive mutex allows reentrant mutex acquisitions
+               // this makes isOk(ip) and checkout(ip) atomic
     std::unordered_map<std::string, std::uint32_t> ipFetchCount_;
+    std::unordered_map<std::string, std::uint32_t> ipRequestCount_;
```

Comment:
This map needs to be cleared at some point. Might be a bit tricky, since `RPCScope` contains a reference a value in the map, and thus we need to be careful not to invalidate the reference. Maybe use `shared_ptrs`?